Add `valohai.distributed` with config parsing

ruksi opened the pull request:
Fixes #87 

TODO:
* [x] add `valohai.distributed` to provide common helpers for reading Valohai distributed task configuration
* [x] add a `valohai.distributed` section to README

Also includes other small improvements here and there.

akx commented on valohai/internals/distributed_config/network.py:
```diff
@@ -0,0 +1,40 @@
+from typing import Any, Dict, List
+
+
+class Network:
```
I'm not sure this needs to be a separate class..?

ruksi replied:
it was a separate well structured object in the JSON hierarchy with different accompanied methods (`primary_local_ip` etc.) so why shouldn't it be?

akx replied:
I don't see why the end user of this library needs to do `x.network.primary_local_ip` when `x.primary_local_ip` is pretty clear too 

Sure, it's a sub-dict in the JSON, but these objects don't need to be an 1:1 mapping to the raw data.

ruksi replied:
yeah, we are not even currently 1:1 as the attributes in JSON `config` object (`distributed.json > .config`) is moved to the root level

should be fine to move one level down as suggested, I doubt the structure will be growing too much? thought we might want to retain it as there it was in the first place in the JSON to group all the networking related things

ruksi commented on valohai/internals/distributed_config/member.py:
```diff
@@ -0,0 +1,40 @@
+        self.member_id = member_id
+        self.network = network
+
+    @property
+    def rank(self) -> int:
+        return member_id_to_rank(self.member_id)
+
+    @property
+    def is_master(self) -> bool:
+        return self.member_id == MASTER_MEMBER_ID
```
yup, many of the `member_id` and `rank` thing would need to change if `member_id` wasn't the `rank` anymore

> think we should just sort the member id strings and use the sorted order as the rank mapping

interesting idea, why not, would work the same be more future anyway? the `int` casting only works in this case but redefined sorting ranking could work just as well in the future

ruksi commented on valohai/distributed.py:
```diff
@@ -0,0 +1,69 @@
+
+    @property
+    def config(self) -> DistributedConfig:
+        if not self._config:
+            with open(self._get_config_path()) as json_file:
+                json_data = json.load(json_file)
+            self._config = DistributedConfig.from_json_data(json_data)
```
yup yup, `FileNotFoundError` is already mentioned in the tests and in the `is_distributed_task` function call

the `FileNotFoundError` already reports which file is missing so figured it would enough for starters

I did start to fiddle with adding a separate exception hierarchy starting with `ValohaiError(Exception)` (as that is how I would usually structure library error handling) and figured it would be too much of change to change all exception classes in the project _included_ in a unrelated PR (not much work, but quite a lot of changes happening at once)

ruksi commented:
* Removed `Network` structure and moved those properties one level up.
* Made ranking work deterministically with non-integer member ids and integers that don't start from zero.

akx commented on valohai/internals/distributed_config/utils.py:
```diff
@@ -0,0 +1,28 @@
+    except ValueError:
+        # if we fail to parse all member ids as an integer, just sort as string
+        id_to_sortable = {member_id: member_id for member_id in member_ids}
+    return {
+        k: index
+        for index, (k, v) in enumerate(
+            sorted(id_to_sortable.items(), key=lambda kv: kv[1])
+        )
+    }
```
Maybe better naming?
```suggestion
    return {
        member_id: index
        for index, (member_id, sort_value) in enumerate(
            sorted(id_to_sortable.items(), key=lambda kv: kv[1])
        )
    }
```

akx commented on valohai/internals/distributed_config/utils.py:
```diff
@@ -0,0 +1,28 @@
+    """Adding ranks to members in-place."""
+    mapping = member_ids_to_rank_mapping([m.member_id for m in members])
+    for m in members:
+        rank = mapping[m.member_id]
+        m.rank = rank
```
```suggestion
        m.rank = mapping[m.member_id]
```

akx commented on valohai/internals/distributed_config/utils.py:
```diff
@@ -0,0 +1,28 @@
+    try:
+        id_to_sortable = {member_id: int(member_id) for member_id in member_ids}
+    except ValueError:
+        # if we fail to parse all member ids as an integer, just sort as string
```
🆗

akx commented on valohai/internals/distributed_config/utils.py:
```diff
@@ -0,0 +1,28 @@
+        m.rank = rank
+
+
+def member_ids_to_rank_mapping(member_ids: Iterable[str]) -> Dict[str, int]:
```
Verb 😁 

```suggestion
def compute_member_id_ranks(member_ids: Iterable[str]) -> Dict[str, int]:
```

ruksi replied:
will fix

akx commented on valohai/distributed.py:
```diff
@@ -0,0 +1,68 @@
+        return self.member(member_id=self.config.member_id)
+
+    def master(self) -> Member:
+        return self.member(member_id=MASTER_MEMBER_ID)
```
I'd like to drop the whole `MASTER_MEMBER_ID` thing, so maybe
```suggestion
        for member in self.members():
            if member.is_master:
                return member
        raise RuntimeError("No master configured")
```

unless this is moved up to the `config`; `.master_member` sounds pretty nice...

ruksi replied:
yeah, we could drop the `MASTER_MEMBER_ID`, will fix

akx commented on valohai/internals/distributed_config/member.py:
```diff
@@ -0,0 +1,60 @@
+        self.exposed_ports = exposed_ports
+        self.local_ips = local_ips
+        self.public_ips = public_ips
+        self.rank = rank  # probably populated later
```
```suggestion
        self.rank = rank  # populated by `DistributedConfig.from_json_data()`
```

akx commented on valohai/distributed.py:
```diff
@@ -0,0 +1,68 @@
+                return member
+        raise RuntimeError(f"No member with id {member_id}")
+
+    def self(self) -> Member:
```
`self.self()` seems so _wrong_, but the only other sane wording I can think of is `.me()`, which isn't better at all really.

Just a comment...

ruksi replied:
yeah, I was thinking the same and we don't __have to__ go with the JSON key naming :thinking: 

will revisit

akx commented on valohai/distributed.py:
```diff
@@ -0,0 +1,68 @@
+        # not a property to mimic `is_running_in_valohai`
+        try:
+            return bool(self.config.group_name)
+        except FileNotFoundError:
+            return False
+        except json.decoder.JSONDecodeError:
+            return False
```
`JSONDecodeError` is re-exported from `json`, you don't generally need to look at the submodules.

```suggestion
        except (FileNotFoundError, json.JSONDecodeError):
            return False
```

ruksi commented on valohai/internals/distributed_config/member.py:
```diff
@@ -0,0 +1,60 @@
+        self.public_ips = public_ips
+        self.rank = rank  # probably populated later
+
+    @property
+    def is_master(self) -> bool:
+        return self.member_id == MASTER_MEMBER_ID
```
could be, will revisit

akx commented on valohai/distributed.py:
```diff
@@ -0,0 +1,68 @@
+        # not a property to mimic `is_running_in_valohai`
+        try:
+            return bool(self.config.group_name)
+        except (FileNotFoundError, json.JSONDecodeError):
+            return False
```
Hey, can we maybe add a bit that makes this function never fail (e.g. if the distributed config changes and we'd be using an old version of valohai-utils)?

```suggestion
        except (FileNotFoundError, json.JSONDecodeError):
            return False
        except Exception as exc:
            warnings.warn(f"Failed to parse distributed config: {exc}")
            return False
```

akx reviewed (changes requested):
One teensy-weensy change request left 🆗

ruksi commented:
never failing try/except added

akx reviewed (approved):
:shipi
